src/Shared/E2ETesting/WaitAssert.cs

```diff
@@ -73,19 +76,48 @@ private static void WaitAssertCore(IWebDriver driver, Action assertion, TimeSpan
             }
             catch (WebDriverTimeoutException)
             {
+                // At this point at least one test failed, so we mark the test as failed. Any assertions after this one
+                // will fail faster. There's a small race condition here between checking the value for TestRunFailed
+                // above and setting it here, but nothing bad can come out of it. Worst case scenario, one or more
+                // tests running concurrently might use the DefaultTimeout in their current assertion, which is fine.
+                TestRunFailed = true;
+
+                var fileId = $"{Guid.NewGuid():N}.png";
+                var screenShotPath = Path.Combine(Path.GetFullPath(E2ETestOptions.Instance.ScreenShotsPath), fileId);
                 var errors = driver.GetBrowserLogs(LogLevel.Severe);
-                if (errors.Count > 0)
-                {
-                    throw new BrowserAssertFailedException(errors, lastException);
-                }
-                else if (lastException != null)
+
+                if (driver is ITakesScreenshot takesScreenshot && E2ETestOptions.Instance.ScreenShotsPath != null)
```

Comment:
This is a bit nitpicky, but I'd consider trying to factor out some of this huge `catch` block (e.g., the code that writes screenshots to disk) into separate methods. It's quite a handful to read and follow the flow as-is, especially with nested `catch` blocks and a local method nested inside one of those.